### PR TITLE
Use ArrayFormatter in Cast Kernel

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #.

# Rationale for this change
 

Using the ArrayFormatter is less code, more consistent, and is ~10% faster as it avoids intermediate string allocations

```
cast i64 to string 512  time:   [10.723 µs 10.728 µs 10.733 µs]
                        change: [-11.208% -11.131% -11.052%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

cast f32 to string 512  time:   [16.210 µs 16.222 µs 16.233 µs]
                        change: [-7.7917% -7.6542% -7.5083%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe
```

# What changes are included in this PR?

# Are there any user-facing changes?

This changes the formatting of Date64 from `1997-05-19 00:00:00` to `1997-05-19T00:00:00`

### CODE REVIEW

File: arrow-cast/src/cast.rs [hunk 1]
```diff
@@ -3457,86 +3247,12 @@ where
     }
 }
 
-/// Helper function to cast from `GenericBinaryArray` to `GenericStringArray`. This function performs
-/// UTF8 validation during casting. For invalid UTF8 value, it could be Null or returning `Err` depending
-/// `CastOptions`.
-fn cast_binary_to_generic_string<I, O>(
```
Comment (tustvold) [inline arrow-cast/src/cast.rs:3463]: I implemented this by converting the offsets and data separately, the performance is the same

File: arrow-cast/src/cast.rs [hunk 2]
```diff
@@ -155,13 +154,12 @@ pub fn can_cast_types(from_type: &DataType, to_type: &DataType) -> bool {
         (_, Boolean) => DataType::is_numeric(from_type) || from_type == &Utf8 || from_type == &LargeUtf8,
         (Boolean, _) => DataType::is_numeric(to_type) || to_type == &Utf8 || to_type == &LargeUtf8,
 
-        (Utf8, LargeUtf8) => true,
```
Comment (alamb) [inline arrow-cast/src/cast.rs:158]: for anyone else reviewing, these case got folded down into the other Utf8 and LargeUtf cases

File: arrow-cast/src/cast.rs [hunk 3]
```diff
@@ -182,11 +181,8 @@ pub fn can_cast_types(from_type: &DataType, to_type: &DataType) -> bool {
             | Time64(TimeUnit::Nanosecond)
             | Timestamp(TimeUnit::Nanosecond, None)
         ) => true,
-        (LargeUtf8, _) => DataType::is_numeric(to_type) && to_type != &Float16,
-        (Timestamp(_, _), Utf8) | (Timestamp(_, _), LargeUtf8) => true,
```
Comment (alamb) [inline arrow-cast/src/cast.rs:186]: What is the rationale for this change? Is the Timestamp nanosecond with no timezone seems to be covered by the cases above, but timestamp with other units and Non null timezone is not (obviously) covered

File: arrow-cast/src/cast.rs [hunk 4]
```diff
@@ -2171,172 +2111,25 @@ where
...
-    iter.into_iter().for_each(|value| {
-        if let Some(value) = value {
-            match as_datetime::<T>(<i64 as From<_>>::from(value)) {
-                Some(dt) => builder.append_value(op(dt)),
-                None => builder.append_null(),
+fn value_to_string<O: OffsetSizeTrait>(
+    array: &dyn Array,
+) -> Result<ArrayRef, ArrowError> {
+    let mut builder = GenericStringBuilder::<O>::new();
+    let options = FormatOptions::default();
```
Comment (alamb) [inline arrow-cast/src/cast.rs:2118]: eventually we could even pass in `FormatOptions` to the cast kernel (definitely not this PR)!

File: arrow-cast/src/cast.rs [hunk 5]
```diff
@@ -2171,172 +2111,25 @@ where
...
+    let formatter = ArrayFormatter::try_new(array, &options)?;
+    let data = array.data();
+    for i in 0..data.len() {
+        match data.is_null(i) {
+            true => builder.append_null(),
+            false => {
+                formatter.value(i).write(&mut builder)?;
+                builder.append_value("");
```
Comment (alamb) [inline arrow-cast/src/cast.rs:2126]:
```suggestion
                // tell the builder the row is finished
                builder.append_value("");
```

File: arrow-cast/src/cast.rs [hunk 6]
```diff
@@ -5521,8 +5237,8 @@ mod tests {
         let b = cast(&array, &DataType::Utf8).unwrap();
         let c = b.as_any().downcast_ref::<StringArray>().unwrap();
         assert_eq!(&DataType::Utf8, c.data_type());
-        assert_eq!("1997-05-19 00:00:00", c.value(0));
-        assert_eq!("2018-12-25 00:00:00", c.value(1));
+        assert_eq!("1997-05-19T00:00:00", c.value(0));
```
Comment (alamb) [inline arrow-cast/src/cast.rs:5240]: Why did these change? This seems like a non trivial change, potentially, for downstream crates, right? Is there some way to get the old behavior back (like maybe passing FormatOptions to the cast kernel) 🤔

**Review by alamb — COMMENTED** [review]
I had some questions, but overall it is pretty sweet to see the performance improve with less code 👍

File: arrow-cast/src/cast.rs (hunk 6 above)
Comment (tustvold) [reply arrow-cast/src/cast.rs:5240]: They change to be RFC3339, given we've made the same change to CSV, JSON and pretty output and this is the last holdout I figured I'd just change it

File: arrow-cast/src/cast.rs (hunk 3 above)
Comment (tustvold) [reply arrow-cast/src/cast.rs:186]: They're rolled into

```
        (_, Utf8 | LargeUtf8) => from_type.is_primitive(),
```

A few lines down

**Review by alamb — APPROVED** [review]
